Extract employee management functionality into Roster component

tcollier opened the pull request:
I've grouped UI elements related to adding/reseting the employee roster and extracted related code into a `<Roster>` component. This allows us to easy move this functionality around as we add to and modify the page. One idea I had was to treat the roster as a separate screen from the group shuffling so there is less clutter on the page for whichever task the admin is doing.

The resulting layout is arguably less attractive. However, rather than invest time in making it pretty at this point, I think we should address the UI later when the functionality has stabilized.

<img width="492" alt="Screen Shot 2020-01-23 at 7 07 19 PM" src="https://user-images.githubusercontent.com/452284/73041805-b0384580-3e13-11ea-808a-c30e009deafd.png">

Commit:
Refactor UI: group roster components

Group components that manage the employee roster in the UI so that
they can easily be extracted into a new React component.

Commit:
Add Roster component

lynnpark commented:
The great thing about frontend dev is that you can worry about functionality AND the UI at the same time!! For the time being, I think it'll be helpful if the Roster component had a title and something visually separating it from the rest of app - maybe a faint gray border? 

I also think it makes sense that the Generate Groups! button is above the Roster component. While you're at it, would you mind making this button larger than the rest? I had always envisioned this button being totally decked out to stand out from the rest.

Commit:
Merge branch 'master' into roster-component

Commit:
Add border and heading to roster component

Commit:
Move generate groups button above roster

Commit:
Increase size of generate groups button

tcollier commented:
Watch out Jony Ive, the new design masters are in town!!

<img width="836" alt="Screen Shot 2020-01-25 at 8 55 35 AM" src="https://user-images.githubusercontent.com/452284/73124445-8b95b800-3f50-11ea-9b71-fdaed5a4e193.png">

lynnpark commented on src/styles.css:
```diff
@@ -24,6 +20,20 @@ button {
+  height: 54px;
+  padding: 12px 27px;
+}
+
```
I know that I originally added the styles to all generic button tags, but it is actually very very bad practice to add global styles to elements 🙀 (shame on u lynn). Also, these classnames are too generic. It's best practice to be verbose with classnames in global css files to avoid accidentally overriding classes!

Ideally, you want to minimize the css that goes into this file, as css files tend to grow fast and become impossible to maintain. Since the Button component is styled dynamically, it would be a good idea to move this into the same or adjacent file as Button. You can pass styles to components as an object with the [style attribute](https://reactjs.org/docs/dom-elements.html#style), although they make it clear that this method is not recommended.

A more elegant solution would be using the CSS-in-JS pattern. Two most popular CSS-in-JS libraries are [styled-components](https://styled-components.com/) and [emotion](https://emotion.sh/docs/introduction). I'm pretty agnostic to either. Here's an interesting [comparison between the two](https://github.com/jsjoeio/styled-components-vs-emotion).

tcollier replied:
Thanks for the detailed explanation! I read the comparison you linked to and the differences sound pretty minimal. `styled-components` has a larger community, whereas `emotion` is lighter weight and more performant. One additional factor is that Apartment List uses `emotion`, so I will work on using that library in this PR.

tcollier replied:
FYI - I had to set the jsx pragma [according to the documentation](https://emotion.sh/docs/css-prop#jsx-pragma) because `create-react-app` does not allow us to configure babel.

Commit:
Add emotion css-in-js library

```
npm i @emotion/core
```

[Emotiono](https://emotion.sh/docs/introduction) is a library designed
for writing css styles with JavaScript. It provides powerful and
predictable style composition in addition to a great developer
experience with features such as source maps, labels, and testing
utilities. Both string and object styles are supported.

Commit:
Style button using emotion css-in-js

Commit:
Don't set className to size

Commit:
Extract common css into const

Commit:
Update snapshots

Commit:
Move roster styling into emotion

Commit:
Update snapshots

lynnpark commented on src/components/Button.js:
```diff
@@ -1,9 +1,54 @@
+      padding: 12px 27px;
+    `
+  }
+
```
I've found that it's best practice to declare variables and functions that do not depend on props outside of the function. This prevents the component from becoming a bloated mess!

lynnpark commented:
One more thing, the spacing around the large button and the Roster components are inconsistent. As a rule thumb, I usually follow [Material Design's guide on spacing](https://material.io/design/layout/spacing-methods.html#spacing). 

The [Material System guide](https://material.io/design/introduction/) is fascinating and a must-read for every front-end developer!

Commit:
Move css const to outside component

Commit:
Format css settings

Commit:
Improve layout

tcollier commented:
I've updated the layout so that the "Generate" button is left aligned with the roster component.

<img width="599" alt="Screen Shot 2020-01-27 at 11 22 32 PM" src="https://user-images.githubusercontent.com/452284/73243285-0ed72980-415c-11ea-9a83-33556b100fd1.png">

Commit:
Update test snapshots

Commit:
Reduce spacing between form elements

tcollier commented:
Sorry, I haven't had the ability to focus on getting this PR merged in. I'm closing for now, but will be happy to reopen when I have more bandwidth.
